packages/react-effect/src/server.tsx

```diff
@@ -27,30 +32,75 @@ export function extract(
     </EffectContext.Provider>
   );
 
-  return (async function perform(): Promise<string> {
+  return (async function perform(index = 0): Promise<string> {
+    const start = Date.now();
     const result = renderFunction(element);
 
     if (manager.finished) {
-      await manager.afterEachPass();
+      const duration = Date.now() - start;
+
+      await manager.afterEachPass({
+        index,
+        finished: manager.finished,
+        renderDuration: duration,
+        resolveDuration: 0,
+      });
+
       if (afterEachPass) {
-        await afterEachPass();
+        await afterEachPass({
+          index,
+          finished: manager.finished,
+          renderDuration: duration,
+          resolveDuration: 0,
+        });
       }
 
       return result;
     } else {
+      const resolveStart = Date.now();
+      const renderDuration = resolveStart - start;
+
       await manager.resolve();
 
-      await manager.betweenEachPass();
+      const resolveDuration = Date.now() - resolveStart;
+
+      await manager.betweenEachPass({
+        index,
+        finished: false,
+        renderDuration,
+        resolveDuration,
+      });
+
       if (betweenEachPass) {
-        await betweenEachPass();
+        await betweenEachPass({
+          index,
+          finished: false,
+          renderDuration,
+          resolveDuration,
+        });
       }
 
-      await manager.afterEachPass();
+      await manager.afterEachPass({
+        index,
+        finished: false,
+        renderDuration,
+        resolveDuration,
+      });
+
       if (afterEachPass) {
-        await afterEachPass();
+        await afterEachPass({
+          index,
+          finished: false,
+          renderDuration,
+          resolveDuration,
+        });
+      }
+
+      if (index + 1 >= maxPasses) {
+        return result;
```

Comment:
So at this point, SSR will just render whatever it has, and unresolved promises will appear client-side with their loading state?

Reply:
Correct. They will get a "mostly server rendered page ™️", and the client will rehydrate with the data it has, and otherwise simply start those queries/ async translation loading/ whatever itself